File: [39] Combination Sum/39.combination-sum.java
import java.util.ArrayList;

/*
 * @lc app=leetcode id=39 lang=java
 *
 * [39] Combination Sum
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        backTracking(candidates, res, new ArrayList<Integer>(), target, 0);
        return res;
    }

    public void backTracking(int[] candidates, List<List<Integer>> res, List<Integer> temp_list, int target, int start) {
        if (target == 0) {
            // Note that we need to new an ArrayList, since temp_list is changing during the rest running time and will become []
            // Explain: https://stackoverflow.com/questions/7080546/add-an-object-to-an-arraylist-and-modify-it-later
            res.add(new ArrayList<>(temp_list));
            return;
        } else if (target < 0) {
            return;
        } else {
            for (int i = start; i < candidates.length; i++) {
                temp_list.add(candidates[i]);
                backTracking(candidates, res, temp_list, target-candidates[i], i);
                temp_list.remove(temp_list.size()-1);
            }
        }
    }
}
// @lc code=end

File: [39] Combination Sum/39.combination-sum.cpp
/*
 * @lc app=leetcode id=39 lang=cpp
 *
 * [39] Combination Sum
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> current;
        backTracking(candidates, res, current, target, 0);
        return res;
    }

    void backTracking(vector<int>& candidates, vector<vector<int>>& res, vector<int>& temp_list, int target, int start) {
        if (target == 0) {
            // no need to new a temp_list, since std::vector<T>::push_back() creates a copy of the argument and stores it in the vector. 
            res.push_back(temp_list);
            return;
        } else if (target < 0) {
            return;
        } else {
            if (start == candidates.size()) return;
            temp_list.push_back(candidates[start]);
            backTracking(candidates, res, temp_list, target-candidates[start], start);
            temp_list.pop_back();
            backTracking(candidates, res, temp_list, target, start+1);
        }
    }
};
// @lc code=end

